fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+
+    def train_by_steps(
+        self, steps: int, current_round: Optional[int] = None
+    ) -> Tuple[Dict[str, float], Dict[str, Scalar]]:
+        """
+        Train locally for the specified number of steps.
+
+        Args:
+            steps (int): The number of steps to train locally.
+
+        Returns:
+            Tuple[Dict[str, float], Dict[str, Scalar]]: The loss and metrics dictionary from the local training.
+                Loss is a dictionary of one or more losses that represent the different components of the loss.
+        """
+        # Need to also set the global model to train mode
+        self.global_model.train()
+        return super().train_by_steps(steps, current_round)
+
+    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> Tuple[TrainingLosses, Dict[str, torch.Tensor]]:
+        """
+        Mechanics of training loop follow from original Ditto implementation: https://github.com/litian96/ditto
+        As in the implementation there, steps of the global and local models are done in tandem and for the same
+        number of steps.
+
+        Args:
+            input (torch.Tensor): input tensor to be run through both the global and local models
+            target (torch.Tensor): target tensor to be used to compute a loss given each models outputs
+
+        Returns:
+            Tuple[TrainingLosses, Dict[str, torch.Tensor]]: Returns relevant loss values from both the global and local
+                model optimization steps. The prediction dictionary contains predictions indexed a "global" and "local"
+                corresponding to predictions from the global and local Ditto models for metric evaluations.
+        """
+
+        # Clear gradients from optimizers if they exist
+        self.optimizers["global"].zero_grad()
+        self.optimizers["local"].zero_grad()
+
+        # Forward pass on both the global and local models
+        preds, features = self.predict(input)
+
+        # Compute all relevant losses
+        losses = self.compute_training_loss(preds, features, target)
+
+        # Take a step with the global model vanilla loss
+        losses.additional_losses["global_loss"].backward()
+        self.optimizers["global"].step()
+
+        # Take a step with the local model using the local loss and Ditto constraint
+        losses.backward["backward"].backward()
+        self.optimizers["local"].step()
+
+        # Return dictionary of predictions where key is used to name respective MetricMeters
+        return losses, preds
+
+    def update_after_step(self, step: int) -> None:
+        if step % self.beta_update_interval == 0:
+            if self.mkmmd_loss_weight and self.init_global_model:
+                # Get the feature distribution of the local and init global features with evaluation mode
+                local_distribution, init_global_distribution = self.update_buffers(self.model, self.init_global_model)
```

Comment:
Maybe I'm missing something down the line, but I doing think we need to compute these distributions if the mkmmd weight is 0? If that's right, then we can tuck these under the if statement below as well to avoid computation when we don't want it to happen.